### PR TITLE
Initial cache migration to new layers system

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Initial cache migration to new layers system, starting from the reactions
store, but settings the foundations to migrate the other stores (roles and
schemes)

#### Extra information about new layers system
The previous Store layers implementation was a more powerful but more complex system, and as far as we weren't using that extra power, we decide to reduce the complexity in order to have a more simple system that can address new functionality with less effort.

The new approach is based on structure embedding, each layer embeds the underneath layer and override the parts that needs some extra behavior. For example, in this case, the cache layer only overrides the ReactionStore, and add the cache there, and relays on the underlying store to resolve the cache misses.

When we initialize the store we instantiate the different layers embeding each one into another. For example, now we have the SqlStore, that is wrapped by the old LayeredStore, that is wrapped by the LocalCacheLayer, which is wrapped by the TimerLayer, so a request to the store now, hits the TimerLayer which measure the time of the request and pass the request to the LocalCacheLayer, the local cache layer try to answer with the cached value, and if it can't, pass the request to the LayeredStore, which (for now) try to answer with the cached value, and if it can't pass it to the SqlStore. Ones the request is resolved it goes back all the way up in the layers.

A super simple example of a layer implementation could be something like:

```go
func NewImportalLayer(baseStore store.Store) InmortalLayer {
    return InmportalLayer{Store: baseStore}
}

type InmortalLayer struct {
    store.Store
}

func (s *InmortalLayer) DropAllTables() {
}
```

This just ignore any DropAllTables call to the store.

#### Ticket
[MM-17832](https://mattermost.atlassian.net/browse/MM-17832)

### CODE REVIEW

**Comment by lindy65** [comment]
Thanks @jespino - looks good to me

Tested the following:
- creating and using custom emoji (jumbo, ordinary and for reactions / in channel header)
- jumbo emoji from center panel and right-hand-side
- adding and removing emoji from already existing emoji reactions
- emoji in channel header

Note: there is an open bug ticket for "large number of emoji reactions causes horizontal scroll" - https://mattermost.atlassian.net/browse/MM-17806 that will be addressed for v5.14

**Comment by lieut-data** [comment]
@jespino, I know we had some discussion in developers meetings, but could you update the description here to explain the architecture of the new approach? It would be a great resource to help explain to the community the ideas behind the change.

**Comment by jespino** [comment]
@lieut-data sure

File: store/local_cache_supplier.go [hunk 1]
```diff
@@ -50,15 +46,13 @@ type ObjectCache interface {
...
-		cluster:       cluster,
+		roleCache:   utils.NewLruWithParams(ROLE_CACHE_SIZE, "Role", ROLE_CACHE_SEC, model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_ROLES),
+		schemeCache: utils.NewLruWithParams(SCHEME_CACHE_SIZE, "Scheme", SCHEME_CACHE_SEC, model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_SCHEMES),
+		metrics:     metrics,
+		cluster:     cluster,
 	}
 
 	if cluster != nil {
-		cluster.RegisterClusterMessageHandler(model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_REACTIONS, supplier.handleClusterInvalidateReaction)
 		cluster.RegisterClusterMessageHandler(model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_ROLES, supplier.handleClusterInvalidateRole)
```
Comment (lieut-data) [inline store/local_cache_supplier.go]: nit: remove extra new line

File: store/localcachelayer/layer.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	localCacheStore := LocalCacheStore{
+		Store:   baseStore,
+		cluster: cluster,
+		metrics: metrics,
+	}
+	localCacheStore.reactionCache = utils.NewLruWithParams(REACTION_CACHE_SIZE, "Reaction", REACTION_CACHE_SEC, model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_REACTIONS)
+	localCacheStore.reaction = LocalCacheReactionStore{ReactionStore: baseStore.Reaction(), rootStore: &localCacheStore}
+
+	if cluster != nil {
+		cluster.RegisterClusterMessageHandler(model.CLUSTER_EVENT_INVALIDATE_CACHE_FOR_REACTIONS, localCacheStore.reaction.handleClusterInvalidateReaction)
```
Comment (lieut-data) [inline store/localcachelayer/layer.go]: nit: spurious extra newline

Comment (jespino) [reply store/localcachelayer/layer.go]: where? I use that line to separate two blocks, the caches initialization and the cluster event message handlers registration.

File: store/local_cache_supplier.go (hunk 1 above)
Comment (jespino) [reply store/local_cache_supplier.go]: Which one? (anyway, I going to remove this whole file in the nexts PRs migrating roles and schemes, and finally removing the whole `LayeredStore`

Comment (lieut-data) [reply store/local_cache_supplier.go]: Hmm, not seeing it anymore. Please disregard :)

File: store/localcachelayer/layer.go (hunk 2 above)
Comment (lieut-data) [reply store/localcachelayer/layer.go]: Hmm, not seeing this one anymore either. Please disregard :)

**Pull request merged**